### PR TITLE
Options to plain json to get all model params

### PR DESCRIPTION
**Pull request by dvpolyakov**
Implemented `OptionsToPlainJson` function that generates plain JSON with all parameters from documentation set in training from `trainingOptions` and `outputFilesOptions`. Unit tests are added. Python interface will be added. Finally it will resolve https://github.com/catboost/catboost/issues/833

I hereby agree to the terms of the CLA available at: [link](https://yandex.ru/legal/cla/?lang=ru).

### CODE REVIEW

File: catboost/libs/options/plain_options_helper.cpp [hunk 1]
```diff
@@ -389,3 +389,92 @@ void NCatboostOptions::PlainJsonToOptions(
 
     trainOptions["flat_params"] = plainOptions;
 }
+
+void NCatboostOptions::OptionsToPlainJson(
```
Comment (annaveronika) [inline catboost/libs/options/plain_options_helper.cpp]: OptionsToPlainJson => ConvertOptionsToPlainJson - function name should start with a verb

File: catboost/libs/options/plain_options_helper.cpp [hunk 2]
```diff
@@ -389,3 +389,92 @@ void NCatboostOptions::PlainJsonToOptions(
...
+    CopyOption(treeOptions, "observations_to_bootstrap", &plainOptionsJson, &seenKeys);
+
+    auto& bootstrapOptions = treeOptions["bootstrap"];
+    CopyOptionWithNewKey(bootstrapOptions, "type", "bootstrap_type", &plainOptionsJson, &seenKeys);
+    CopyOption(bootstrapOptions, "bagging_temperature", &plainOptionsJson, &seenKeys);
+    CopyOption(bootstrapOptions, "subsample", &plainOptionsJson, &seenKeys);
+    CopyOption(bootstrapOptions, "mvs_head_fraction", &plainOptionsJson, &seenKeys);
+
+    //feature evaluation options
+    auto& modelBasedEvalOptions = options["model_based_eval_options"];
```
Comment (annaveronika) [inline catboost/libs/options/plain_options_helper.cpp]: These options should not be present here, they are only for eval.

File: catboost/libs/options/plain_options_helper.cpp [hunk 3]
```diff
@@ -389,3 +389,92 @@ void NCatboostOptions::PlainJsonToOptions(
...
+    CopyOption(boostingOptionsRef, "learning_rate", &plainOptionsJson, &seenKeys);
+    CopyOption(boostingOptionsRef, "fold_len_multiplier", &plainOptionsJson, &seenKeys);
+    CopyOption(boostingOptionsRef, "approx_on_full_history", &plainOptionsJson, &seenKeys);
+    CopyOption(boostingOptionsRef, "fold_permutation_block", &plainOptionsJson, &seenKeys);
+    CopyOption(boostingOptionsRef, "min_fold_size", &plainOptionsJson, &seenKeys);
+    CopyOption(boostingOptionsRef, "permutation_count", &plainOptionsJson, &seenKeys);
+    CopyOption(boostingOptionsRef, "boosting_type", &plainOptionsJson, &seenKeys);
+    CopyOption(boostingOptionsRef, "data_partition", &plainOptionsJson, &seenKeys);
+
+    auto& odConfig = boostingOptionsRef["od_config"];
```
Comment (annaveronika) [inline catboost/libs/options/plain_options_helper.cpp]: Please remove all the options with empty values.

File: catboost/libs/options/plain_options_helper.cpp [hunk 4]
```diff
@@ -389,3 +389,92 @@ void NCatboostOptions::PlainJsonToOptions(
...
+    CopyOption(treeOptions, "random_strength", &plainOptionsJson, &seenKeys);
+    CopyOption(treeOptions, "leaf_estimation_method", &plainOptionsJson, &seenKeys);
+    CopyOption(treeOptions, "grow_policy", &plainOptionsJson, &seenKeys);
+    CopyOption(treeOptions, "max_leaves", &plainOptionsJson, &seenKeys);
+    CopyOption(treeOptions, "min_data_in_leaf", &plainOptionsJson, &seenKeys);
+    CopyOption(treeOptions, "score_function", &plainOptionsJson, &seenKeys);
+    CopyOption(treeOptions, "fold_size_loss_normalization", &plainOptionsJson, &seenKeys);
+    CopyOption(treeOptions, "add_ridge_penalty_to_loss_function", &plainOptionsJson, &seenKeys);
+    CopyOption(treeOptions, "sampling_frequency", &plainOptionsJson, &seenKeys);
+    CopyOption(treeOptions, "dev_max_ctr_complexity_for_border_cache", &plainOptionsJson, &seenKeys);
```
Comment (annaveronika) [inline catboost/libs/options/plain_options_helper.cpp]: please remove all dev_* parameters

File: catboost/libs/options/plain_options_helper.cpp [hunk 5]
```diff
@@ -389,3 +389,92 @@ void NCatboostOptions::PlainJsonToOptions(
...
+    auto& bootstrapOptions = treeOptions["bootstrap"];
+    CopyOptionWithNewKey(bootstrapOptions, "type", "bootstrap_type", &plainOptionsJson, &seenKeys);
+    CopyOption(bootstrapOptions, "bagging_temperature", &plainOptionsJson, &seenKeys);
+    CopyOption(bootstrapOptions, "subsample", &plainOptionsJson, &seenKeys);
+    CopyOption(bootstrapOptions, "mvs_head_fraction", &plainOptionsJson, &seenKeys);
+
+    //feature evaluation options
+    auto& modelBasedEvalOptions = options["model_based_eval_options"];
+
+    CopyOption(modelBasedEvalOptions, "features_to_evaluate", &plainOptionsJson, &seenKeys);
```
Comment (annaveronika) [inline catboost/libs/options/plain_options_helper.cpp]: I don't see ctr options and ignored features here

File: result.log [hunk 6]
```diff
@@ -0,0 +1 @@
+/home/daniil/.ya/build/symres/1abe115d91aa2a026292ca981c1084c0/result.log
```
Comment (annaveronika) [inline result.log]: i don't think we need logs here

File: catboost/libs/options/plain_options_helper.cpp [hunk 7]
```diff
@@ -389,3 +389,92 @@ void NCatboostOptions::PlainJsonToOptions(
 
     trainOptions["flat_params"] = plainOptions;
 }
+
+void NCatboostOptions::OptionsToPlainJson(
+        const NJson::TJsonValue& options,
+        const NJson::TJsonValue& outputOptions,
+        NJson::TJsonValue* plainOptions)
+{
+    TSet<TString> seenKeys;
```
Comment (annaveronika) [inline catboost/libs/options/plain_options_helper.cpp]: You need to add a check that everything from json is put to seenKeys or explicitly ignored. To do that you can remove seen elements from options and then check that nothing has left.

File: catboost/libs/algo/full_model_saver.cpp [hunk 8]
```diff
@@ -266,6 +268,11 @@ namespace NCB {
             NJson::TJsonValue jsonOptions(NJson::EJsonValueType::JSON_MAP);
             Options.Save(&jsonOptions);
             dstModel->ModelInfo["params"] = ToString(jsonOptions);
+
+            NJson::TJsonValue jsonOutputOptions(NJson::EJsonValueType::JSON_MAP);
```
Comment (annaveronika) [inline catboost/libs/algo/full_model_saver.cpp]: why is this needed?

File: catboost/libs/options/plain_options_helper.cpp [hunk 9]
```diff
@@ -388,4 +388,203 @@ void NCatboostOptions::PlainJsonToOptions(
...
+
+    NJson::TJsonValue& plainOptionsJson = *plainOptions;
+    plainOptionsJson.SetType(NJson::JSON_MAP);
+
+    // loss reverse mapping is not implemented yet
+//    auto& lossFunctionRef = trainOptions["loss_function"];
```
Comment (annaveronika) [inline catboost/libs/options/plain_options_helper.cpp]: please remove the commented out code

File: catboost/libs/options/cat_feature_options.cpp [hunk 10]
```diff
@@ -324,3 +326,48 @@ bool NCatboostOptions::CtrsNeedTargetData(const NCatboostOptions::TCatFeaturePar
 
     return ctrsNeedTargetData;
 }
+
+TString NCatboostOptions::RemapAndConcatenateCtrOptions(const NJson::TJsonValue& options)
+{
+    TString ctr_type_string_concat = ToString(options["ctr_type"]);
```
Comment (annaveronika) [inline catboost/libs/options/cat_feature_options.cpp]: camelCase for C++, please read the styleguilde

File: catboost/libs/options/cat_feature_options.cpp [hunk 11]
```diff
@@ -324,3 +326,48 @@ bool NCatboostOptions::CtrsNeedTargetData(const NCatboostOptions::TCatFeaturePar
 
     return ctrsNeedTargetData;
 }
+
+TString NCatboostOptions::RemapAndConcatenateCtrOptions(const NJson::TJsonValue& options)
```
Comment (annaveronika) [inline catboost/libs/options/cat_feature_options.cpp]: RemapAndConcatenateCtrOptions => BuildCtrOptionsDescription

File: catboost/libs/options/cat_feature_options.cpp [hunk 12]
```diff
@@ -324,3 +326,48 @@ bool NCatboostOptions::CtrsNeedTargetData(const NCatboostOptions::TCatFeaturePar
...
+    if (options["target_binarization"].Has("border_type")) {
+        auto target_border_type_stripped = StripString(ToString(options["target_binarization"]["border_type"]),
+                                                       EqualsStripAdapter('"'));
+        ctr_type_string_concat =
+                ctr_type_string_concat + ":TargetBorderType=" + target_border_type_stripped;
+    }
+
+    const NJson::TJsonValue& priorDescriptions = options["priors"];
+    if (priorDescriptions.IsArray()) {
+        for (const auto &prior : priorDescriptions.GetArraySafe()) {
```
Comment (annaveronika) [inline catboost/libs/options/cat_feature_options.cpp]: const auto& see styleguilde

File: catboost/libs/options/loss_description.cpp [hunk 13]
```diff
@@ -266,3 +269,21 @@ NJson::TJsonValue LossDescriptionToJson(const TStringBuf lossDescription) {
     }
     return descriptionJson;
 }
+
+TString RemapLossOrMetricsOptions(const NJson::TJsonValue& options) {
```
Comment (annaveronika) [inline catboost/libs/options/loss_description.cpp]: BuildMetricOptionDescription

File: catboost/libs/options/loss_description.cpp [hunk 14]
```diff
@@ -266,3 +269,21 @@ NJson::TJsonValue LossDescriptionToJson(const TStringBuf lossDescription) {
...
+    TString ParamType = StripString(ToString(options["type"]), EqualsStripAdapter('"'));
+    if (!options["params"].GetMap().empty()) {
+        ParamType = ParamType + ":";
+    }
+    for (auto& elem : options["params"].GetMap()) {
```
Comment (annaveronika) [inline catboost/libs/options/loss_description.cpp]: const auto&

File: catboost/libs/options/loss_description.h [hunk 15]
```diff
@@ -70,3 +70,5 @@ TMap<TString, TString> ParseHintsDescription(TStringBuf hintsDescription);
 TString MakeHintsDescription(const TMap<TString, TString>& hints);
 
 NJson::TJsonValue LossDescriptionToJson(TStringBuf lossDescription);
+TString RemapLossOrMetricsOptions(const NJson::TJsonValue&);
```
Comment (annaveronika) [inline catboost/libs/options/loss_description.h]: please add parameter name here

File: catboost/libs/options/plain_options_helper.cpp [hunk 16]
```diff
@@ -388,4 +439,141 @@ void NCatboostOptions::PlainJsonToOptions(
...
+
+    plainOptionsJson["eval_metric"] = RemapLossOrMetricsOptions(options["metrics"]["eval_metric"]);
+    DeleteSeenOption(&OptionsCopy["metrics"], "eval_metric");
+
+    auto& result = plainOptionsJson["custom_metric"] = NJson::TJsonValue(NJson::JSON_ARRAY);
+    for (auto& metric : options["metrics"]["custom_metrics"].GetArray()) {
+        result.AppendValue(RemapLossOrMetricsOptions(metric));
+    }
+    DeleteSeenOption(&OptionsCopy["metrics"], "custom_metrics");
+    DeleteSeenOption(&OptionsCopy["metrics"], "objective_metric");
```
Comment (annaveronika) [inline catboost/libs/options/plain_options_helper.cpp]: objective_metric - это метрика, которая соответствует нашему лоссу, которая потом устанавливается в eval_metric, если он не установлен, либо если он установлен, то просто считается и выводится в логи с ошибками, а также используется для LossFunctionChange fstr-ов.
This means that this value cannot be blindly deleted, it should also be in options.

**Comment by annaveronika** [comment]
Could you please also add:
if (od_config.type == None) od_config.{other_options} should not be included.
if ("node_type":"SingleHost") "file_with_hosts" and "node_port" should not be included.
if (no cat features in model) "cat_feature_params" should not be included.

**Comment by annaveronika** [comment]
Please fix styleguilde in the latest version.

**Comment by annaveronika** [comment]
few styleguilde things to check:
- naming in C++ (camelCase starting with low letter for function parameters and local variables)
- space with &: is together with previous word like that for (const auto& x : y)
- no more than single consecutive empty line inside a function, and no more than two between the functions
- no abbreviations at all (exceptions: idx, len, elem, dst, src)

**Comment by annaveronika** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@dvpolyakov, please agree to the CLA (https://github.com/catboost/catboost/blob/master/CONTRIBUTING.md#provide-contributions)

**Comment by arcadia-devtools** [comment]
CLA already signed

**Comment by annaveronika** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@annaveronika, internal review request created: [841568](https://a.yandex-team.ru/review/841568)

**Comment by arcadia-devtools** [comment]
Check [841568](https://a.yandex-team.ru/review/841568) failed: 17 test(s) broken

**Comment by annaveronika** [comment]
@dvpolyakov Could you please resolve conflicts?

**Comment by annaveronika** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@annaveronika, internal review request updated: [841568](https://a.yandex-team.ru/review/841568)

**Comment by arcadia-devtools** [comment]
Check [841568](https://a.yandex-team.ru/review/841568) failed: 265 build(s) broken, 182 test(s) broken

**Comment by annaveronika** [comment]
moved discussion to https://github.com/catboost/catboost/pull/883
